### PR TITLE
2fa integration ui

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: cmd/keymasterd/main.go [hunk 1]
```diff
@@ -311,6 +362,13 @@ func (state *RuntimeState) checkAuth(w http.ResponseWriter, r *http.Request) (st
 		authCookie = cookie
 	}
 	if authCookie == nil {
+
+		if (AuthTypePassword & requiredAuthType) == 0 {
+			state.writeFailureResponse(w, r, http.StatusUnauthorized, "")
+			err := errors.New("Insufficeint Auth Level passwd")
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: This is opaque to the user. What does this mean?

File: cmd/keymasterd/main.go [hunk 2]
```diff
@@ -354,9 +412,34 @@ func (state *RuntimeState) checkAuth(w http.ResponseWriter, r *http.Request) (st
 		return "", AuthTypeNone, err
 
 	}
+	if (info.AuthType & requiredAuthType) == 0 {
+		state.writeFailureResponse(w, r, http.StatusUnauthorized, "")
+		err := errors.New("Insufficeint Auth Level")
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: Also not clear.

File: cmd/keymasterd/main.go [hunk 3]
```diff
@@ -354,9 +412,34 @@ func (state *RuntimeState) checkAuth(w http.ResponseWriter, r *http.Request) (st
...
+		return "", info.AuthType, err
+	}
 	return info.Username, info.AuthType, nil
 }
 
+func (state *RuntimeState) getRequiredWebUIAuthLevel() int {
+	AuthLevel := 0
+	for _, webUIPref := range state.Config.Base.AllowedAuthBackendsForWebUI {
+		if webUIPref == proto.AuthTypePassword {
+			AuthLevel = AuthLevel | AuthTypePassword
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: FYI: you can do:
`    AuthLevel |= AuthTypePassword`

File: cmd/keymasterd/main.go [hunk 4]
```diff
@@ -836,6 +919,7 @@ func (state *RuntimeState) loginHandler(w http.ResponseWriter, r *http.Request)
 	}
 
 	// AUTHN has passed
+	//logger.Printf("Valid passwd AUTH login for %s", username)
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: Please don't comment out, instead use `logger.Debugf(...)`

File: cmd/keymasterd/main.go [hunk 5]
```diff
@@ -995,19 +1085,23 @@ func (state *RuntimeState) VIPAuthHandler(w http.ResponseWriter, r *http.Request
...
 	if !valid {
+		logger.Printf("Invalid OTP value login for %s", authUser)
 		// TODO if client is html then do a redirect back to vipLoginPage
 		state.writeFailureResponse(w, r, http.StatusUnauthorized, "")
-		logger.Printf("Invalid OTP value login for %s", authUser)
 		return
 
 	}
+
+	// OTP check has been successful
```
Comment (rgooch) [inline cmd/keymasterd/main.go]: s/has been/was/

File: cmd/keymasterd/main.go (hunk 1 above)
Comment (cviecco) [reply cmd/keymasterd/main.go]: To the use they get a 401 error page. For the logs it tell us that there was an a log-in attempt, but the url requested more than a password for login (and there were no cookies in the call).

File: cmd/keymasterd/main.go (hunk 2 above)
Comment (cviecco) [reply cmd/keymasterd/main.go]: similar as above a call was made, but the authorization level was not eough given the current user cookies.

File: cmd/keymasterd/main.go (hunk 3 above)
Comment (cviecco) [reply cmd/keymasterd/main.go]: and will change all of them

**Pull request merged**